File: src/services/references/types/findRefsTypes.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

type ref_kind =
  | PropertyDefinition
  | PropertyAccess
  | Local

let string_of_ref_kind = function
  | PropertyDefinition -> "PropertyDefinition"
  | PropertyAccess -> "PropertyAccess"
  | Local -> "Local"

type single_ref = ref_kind * Loc.t

type find_refs_found = single_ref list

type find_refs_ok = find_refs_found option

type kind =
  | FindReferences
  | Rename

type request = {
  def_info: Get_def_types.def_info;
  kind: kind;
}

let empty_request = { def_info = Get_def_types.NoDefinition; kind = FindReferences }

File: rust_port/crates/flow_services_references/src/find_refs_types.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

//! Port of `services/references/types/findRefsTypes.ml`

use flow_parser::loc::Loc;
use flow_services_get_def::get_def_types::DefInfo;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum RefKind {
    PropertyDefinition,
    PropertyAccess,
    Local,
}

pub fn string_of_ref_kind(kind: RefKind) -> &'static str {
    match kind {
        RefKind::PropertyDefinition => "PropertyDefinition",
        RefKind::PropertyAccess => "PropertyAccess",
        RefKind::Local => "Local",
    }
}

pub type SingleRef = (RefKind, Loc);

pub type FindRefsFound = Vec<SingleRef>;

#[derive(Debug, Clone)]
pub enum FindRefsOk {
    FoundReferences(FindRefsFound),
    NoDefinition(Option<String>),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Kind {
    FindReferences,
    Rename,
}

#[derive(Debug, Clone)]
pub struct Request {
    pub def_info: DefInfo,
    pub kind: Kind,
}

pub fn empty_request() -> Request {
    Request {
        def_info: DefInfo::NoDefinition(None),
        kind: Kind::FindReferences,
    }
}
